pkg/agent/cniserver/pod_configuration.go

```diff
@@ -666,3 +631,129 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+func (pc *podConfigurator) disconnectInterfaceFromOVS(containerConfig *interfacestore.InterfaceConfig) error {
+	containerID := containerConfig.ContainerID
+	klog.V(2).Infof("Deleting Openflow entries for container %s", containerID)
+	if err := pc.ofClient.UninstallPodFlows(containerConfig.InterfaceName); err != nil {
+		return fmt.Errorf("failed to delete Openflow entries for container %s: %v", containerID, err)
+	}
+	klog.V(2).Infof("Deleting OVS port %s for container %s", containerConfig.PortUUID, containerID)
+	// TODO: handle error and introduce garbage collection for failure on deletion
+	if err := pc.ovsBridgeClient.DeletePort(containerConfig.PortUUID); err != nil {
+		return fmt.Errorf("failed to delete OVS port for container %s: %v", containerID, err)
+	}
+	// Remove container configuration from cache.
+	pc.ifaceStore.DeleteInterface(containerConfig)
+	klog.Infof("Removed interfaces for container %s", containerID)
+	return nil
+}
+
+// connectInterceptedInterface connects intercepted interface to ovs br-int.
+func (pc *podConfigurator) connectInterceptedInterface(
+	podName string,
+	podNameSpace string,
+	containerID string,
+	containerNetNS string,
+	containerIFDev string,
+	containerIPs []*current.IPConfig,
+) error {
+	netNS, err := ns.GetNS(containerNetNS)
+	if err != nil {
+		return fmt.Errorf("failed to open netns %s: %v", containerNetNS, err)
+	}
+	defer netNS.Close()
+
+	containerIface := &current.Interface{}
+	intf, err := util.GetNSDevInterface(containerNetNS, containerIFDev)
+	if err != nil {
+		return fmt.Errorf("connectInterceptedInterface failed to get veth info: %w", err)
+	}
+	containerIface.Name = containerIFDev
+	containerIface.Sandbox = netNS.Path()
+	containerIface.Mac = intf.HardwareAddr.String()
+
+	// setup dev in host ns.
+	hostIface := &current.Interface{}
+	intf, br, err := util.GetNSPeerDevBridge(containerNetNS, containerIFDev)
+	if err != nil {
+		return fmt.Errorf("connectInterceptedInterface failed to get veth peer info: %w", err)
+	}
+	if len(br) > 0 {
+		return fmt.Errorf("connectInterceptedInterface: does not expect device %s attached to bridge", intf.Name)
+	}
+	hostIface.Name = intf.Name
+	hostIface.Mac = intf.HardwareAddr.String()
+	if err = pc.routeClient.MigrateRoutesToGw(intf.Name); err != nil {
+		return fmt.Errorf("connectInterceptedInterface failed to migrate: %w", err)
+	}
+	return pc.connectInterfaceToOVS(podName, podNameSpace, containerID, hostIface,
+		containerIface, containerIPs)
+}
+
+// disconnectInterceptedInterface disconnects intercepted interface from ovs br-int.
+func (pc *podConfigurator) disconnectInterceptedInterface(podName, podNamespace, containerID string) error {
+	containerConfig, found := pc.ifaceStore.GetContainerInterface(podName, podNamespace)
+	if !found {
+		klog.V(2).Infof("Did not find the port for container %s in local cache", containerID)
+		return nil
+	}
+	if err := pc.routeClient.UnMigrateRoutesFromGw(&net.IPNet{
```

Comment:
Is it needed or we can just remove routes?

Reply:
UnMigrateRouteFromGw does some error checking, also good to encapsulate.